File: src/main/java/com/lullaby/nutritioncalculatorapi/domain/Ingredient.java
package com.lullaby.nutritioncalculatorapi.domain;

import com.lullaby.nutritioncalculatorapi.common.jpa.BaseEntity;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Getter
@EntityListeners(AuditingEntityListener.class)
@Entity
public class Ingredient extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "ingredient_id")
    private Long id;
    private String name;
    private String memo;
    private Integer unitPrice;
    private Double calories;
    private Double protein;
    private Double fat;
    private Double carbohydrates;
    private Double sugar;
    private Double fiber;
    private Double caffeine;
    private Double saturatedFat;
    @Enumerated(EnumType.STRING)
    private IngredientCategory category;

    protected Ingredient() {}
    public Ingredient(String name, String memo, Integer unitPrice, Double calories, Double protein, Double fat, Double carbohydrates, Double sugar, Double fiber, Double caffeine, Double saturatedFat, IngredientCategory category) {
        setName(name);
        setUnitPrice(unitPrice);
        setCalories(calories);
        setProtein(protein);
        setFat(fat);
        setCarbohydrates(carbohydrates);
        setSugar(sugar);
        setFiber(fiber);
        setCaffeine(caffeine);
        setSaturatedFat(saturatedFat);
        setMemo(memo);
        setCategory(category);
    }
    public void setName(String name) {
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("이름은 필수입니다.");
        }
        this.name = name;
    }

    public void setUnitPrice(Integer unitPrice) {
        if (unitPrice == null || unitPrice < 0) {
            throw new IllegalArgumentException("단위 가격은 0이상 이어야 합니다.");
        }
        this.unitPrice = unitPrice;
    }

    public void setCalories(Double calories) {
        if (calories == null || calories < 0) {
            throw new IllegalArgumentException("칼로리는 0이상 이어야 합니다.");
        }
        this.calories = calories;
    }

    public void setProtein(Double protein) {
        if (protein == null || protein < 0) {
            throw new IllegalArgumentException("단백질은 0이상 이어야 합니다.");
        }
        this.protein = protein;
    }

    public void setFat(Double fat) {
        if (fat == null || fat < 0) {
            throw new IllegalArgumentException("지방은 0이상 이어야 합니다.");
        }
        this.fat = fat;
    }

    public void setCarbohydrates(Double carbohydrates) {
        if (carbohydrates == null || carbohydrates < 0) {
            throw new IllegalArgumentException("탄수화물은 0이상 이어야 합니다.");
        }
        this.carbohydrates = carbohydrates;
    }

    public void setSugar(Double sugar) {
        if (sugar == null || sugar < 0) {
            throw new IllegalArgumentException("당류는 0이상 이어야 합니다.");
        }
        this.sugar = sugar;
    }

    public void setFiber(Double fiber) {
        if (fiber == null || fiber < 0) {
            throw new IllegalArgumentException("식이섬유는 0이상 이어야 합니다.");
        }
        this.fiber = fiber;
    }

    public void setCaffeine(Double caffeine) {
        if (caffeine == null || caffeine < 0) {
            throw new IllegalArgumentException("카페인은 0이상 이어야 합니다.");
        }
        this.caffeine = caffeine;
    }

    public void setSaturatedFat(Double saturatedFat) {
        if (saturatedFat == null || saturatedFat < 0) {
            throw new IllegalArgumentException("포화지방은 0이상 이어야 합니다.");
        }
        this.saturatedFat = saturatedFat;
    }

    public void setMemo(String memo) {
        this.memo = memo;
    }

    public void setCategory(IngredientCategory category) {
        if (category == null) {
            throw new IllegalArgumentException("카테고리는 필수입니다.");
        }
        this.category = category;
    }
}

File: src-electron/ingredient/ingredient.ts
export const IngredientCategory = {
  FRESH: {
    name: 'fresh',
    label: '신선 식품',
  },
  FRUITS: {
    name: 'fruits',
    label: '과일 & 채소',
  },
  FLOUR: {
    name: 'flour',
    label: '가루류 & 곡류',
  },
  COFFEE: {
    name: 'coffee',
    label: '커피 원두 & 차',
  },
  TOPPING: {
    name: 'topping',
    label: '토핑 재료',
  },
  ADDITIVES: {
    name: 'additives',
    label: '식품 첨가물',
  },
  PACKAGING: {
    name: 'packaging',
    label: '포장 용품',
  },
}

export type IngredientCategory = typeof IngredientCategory[keyof typeof IngredientCategory];
export const valueOfIngredientCategory = (name: string): IngredientCategory => {
  switch (name.toUpperCase()) {
    case 'FRESH':
      return IngredientCategory.FRESH;
    case 'FRUITS':
      return IngredientCategory.FRUITS;
    case 'FLOUR':
      return IngredientCategory.FLOUR;
    case 'COFFEE':
      return IngredientCategory.COFFEE;
    case 'TOPPING':
      return IngredientCategory.TOPPING;
    case 'ADDITIVES':
      return IngredientCategory.ADDITIVES;
    case 'PACKAGING':
      return IngredientCategory.PACKAGING;
    default:
      throw new Error(`식자재 카테고리를 찾을 수 없습니다. [name = ${name}]`);
  }
}


export default class Ingredient {
  private id?: number;
  private name = '';
  private memo?: string = '';
  private category: IngredientCategory = IngredientCategory.FRESH;
  private calories = 0;
  private unitPrice = 0;
  private carbohydrates = 0;
  private sugars = 0;
  private protein = 0;
  private caffeine = 0;
  private fat = 0;
  private saturatedFat = 0;
  private createdAt?: Date;
  private updatedAt?: Date;

  public getId(): number | undefined {
    return this.id;
  }

  public setId(value: number) {
    if (!value || value < 1) {
      throw new Error('식자재 ID는 1보다 작을 수 없습니다.');
    }

    this.id = value;
  }

  public getName(): string {
    return this.name;
  }

  public setName(value: string) {
    if (!value) {
      throw new Error('이름을 입력해 주세요.');
    }
    this.name = value;
  }

  public getMemo(): string {
    return this.memo || '';
  }

  public setMemo(value: string) {
    this.memo = value;
  }

  public getCategory(): IngredientCategory {
    return this.category;
  }

  public setCategory(value: IngredientCategory) {
    this.category = value;
  }

  public getCalories(): number {
    return this.calories;
  }

  public setCalories(value: number) {
    if (value < 0) {
      throw new Error('칼로리는 0보다 작을 수 없습니다.');
    }
    this.calories = Number(value);
  }

  public getUnitPrice(): number {
    return this.unitPrice;
  }

  public setUnitPrice(value: number) {
    if (value < 0) {
      throw new Error('단위 가격은 0보다 작을 수 없습니다.');
    }
    this.unitPrice = Number(value);
  }

  public getCarbohydrates(): number {
    return this.carbohydrates;
  }

  public setCarbohydrates(value: number) {
    if (value < 0) {
      throw new Error('탄수화물은 0보다 작을 수 없습니다.');
    }
    this.carbohydrates = Number(value);
  }

  public getSugars(): number {
    return this.sugars;
  }

  public setSugars(value: number) {
    if (value < 0) {
      throw new Error('당류는 0보다 작을 수 없습니다.');
    }
    this.sugars = Number(value);
  }

  public getProtein(): number {
    return this.protein;
  }

  public setProtein(value: number) {
    if (value < 0) {
      throw new Error('단백질은 0보다 작을 수 없습니다.');
    }
    this.protein = Number(value);
  }

  public getCaffeine(): number {
    return this.caffeine;
  }

  public setCaffeine(value: number) {
    if (value < 0) {
      throw new Error('카페인은 0보다 작을 수 없습니다.');
    }
    this.caffeine = Number(value);
  }

  public getFat(): number {
    return this.fat;
  }

  public setFat(value: number) {
    if (value < 0) {
      throw new Error('지방은 0보다 작을 수 없습니다.');
    }
    this.fat = Number(value);
  }

  public getSaturatedFat(): number {
    return this.saturatedFat;
  }

  public setSaturatedFat(value: number) {
    if (value < 0) {
      throw new Error('포화지방산은 0보다 작을 수 없습니다.');
    }
    this.saturatedFat = Number(value);
  }

  public getCreatedAt(): Date | undefined {
    return this.createdAt;
  }

  public setCreatedAt(value: Date) {
    if (!this.createdAt) {
      this.createdAt = value;
    }
  }

  public getUpdatedAt(): Date | undefined{
    return this.updatedAt;
  }

  public setUpdatedAt(value: Date) {
    this.updatedAt = value;
  }

}
